### PR TITLE
Preferences: join EQ and crossfader settings into new Mixer page

### LINKED ISSUE #9401: Combine EQ and XFader to Mixer page in Preferencs
Reported by: **[kazakore](https://github.com/kazakore)**
Date: 2018-08-09T09:54:01Z
Status: Confirmed
Importance: Wishlist
Launchpad Issue: [lp1786211](https://bugs.launchpad.net/bugs/1786211)
Tags: preferences

---

(Wishlist/feature change)

I feel it may help people find certain settings if the preferences for everything which could be considered part of the Mixer were on a single page in Preferences. This has come to my attention due to a user struggling to find the Quick Effects options after losing the control knob. [1]

I believe currently this would mean a merge of the Equalisers and the Crossfader page into a single page. Although this will increase the size of the page it would still be smaller than some others (eg Sound Hardware) and the Preferences is already vertically scrollable so this should not be a huge issue IMO.

Assumedly a new Preference page will be added at some point for Effects (managing saved FXChains and visible 3rd party effects if LV2 support comes in) so whether Quick Effects belongs here on in the Mixer section may be worth a discussion. Or maybe the Effects stuff with be done through a sub pane/window of its own rather than Preferences.... I don't know but may be worth thinking about in connection to this.

[1] https://www.mixxx.org/forums/viewtopic.php?f=3&t=12032

### PR DESCRIPTION
**Pull request by ronso0**
Now all mixer settings are in one place.
Plus some small tweaks, incl. the Main EQ fix from #11271 

Closes #9401

### CODE REVIEW

File: src/preferences/dialog/dlgprefmixer.h [hunk 1]
```diff
@@ -73,9 +76,23 @@ class DlgPrefEQ : public DlgPreferencePage, public Ui::DlgPrefEQDlg  {
...
+    QGraphicsScene* m_pxfScene;
+
+    // X-fader values
+    double m_xFaderMode, m_transform, m_cal;
+
+    PollingControlProxy m_mode;
+    PollingControlProxy m_curve;
+    PollingControlProxy m_calibration;
+    PollingControlProxy m_reverse;
+    PollingControlProxy m_crossfader;
```
Comment (Swiftb0y) [inline src/preferences/dialog/dlgprefmixer.h:88]: does somebody know why these are `PollingControlProxy`s?

File: src/preferences/dialog/dlgprefmixer.cpp [hunk 2]
```diff
@@ -635,7 +731,16 @@ void DlgPrefEQ::slotApply() {
...
 
 // supposed to set the widgets to match internal state
-void DlgPrefEQ::slotUpdate() {
+void DlgPrefMixer::slotUpdate() {
+    if (radioButtonAdditive->isChecked()) {
+        m_xFaderMode = MIXXX_XFADER_ADDITIVE;
+    }
+    if (radioButtonConstantPower->isChecked()) {
+        m_xFaderMode = MIXXX_XFADER_CONSTPWR;
+    }
```
Comment (Swiftb0y) [inline src/preferences/dialog/dlgprefmixer.cpp:742]: shouldn't this rather be put into `slotApply`?

File: src/preferences/dialog/dlgprefmixer.cpp [hunk 3]
```diff
@@ -57,6 +71,26 @@ DlgPrefEQ::DlgPrefEQ(
...
+
+    loadSettings();
+
+    connect(SliderXFader,
+            QOverload<int>::of(&QSlider::valueChanged),
+            this,
+            &DlgPrefMixer::slotUpdateXFader);
+    connect(SliderXFader, &QSlider::sliderMoved, this, &DlgPrefMixer::slotUpdateXFader);
+    connect(SliderXFader, &QSlider::sliderReleased, this, &DlgPrefMixer::slotUpdateXFader);
+    connect(SliderXFader, &QSlider::sliderReleased, this, &DlgPrefMixer::slotApply);
```
Comment (Swiftb0y) [inline src/preferences/dialog/dlgprefmixer.cpp:87]: I doubt its a good UX if simply releasing the slider instantly applies the settings.

**Review by Swiftb0y — COMMENTED** [review]
I'm aware that these are mostly suggestions for improvements of the code you just copied from the old class... Also I can't really comment on the UI XML stuff.

File: src/preferences/dialog/dlgprefmixer.h (hunk 1 above)
Comment (ronso0) [reply src/preferences/dialog/dlgprefmixer.h:88]: because they don't require signal connections, just read and set on demand in update/apply slots

**Comment by ronso0** [comment]
@Swiftb0y Thanks for taking a look, though there are no relevant functional changes, except the Main EQ bugfix.
I'd appreciate if we could get this merged quickly.

This entire page is a bit messy and sub-optimal in terms of UX. I already touched it in #4667 and I don't plan to do that again, at least not in this PR. Though, any notes are helpful in case someone want's to pick that up again.


Any objections to use the crossfader icon for this Mixer page in the preferences sidebar?

File: src/preferences/dialog/dlgprefmixer.cpp (hunk 3 above)
Comment (ronso0) [reply src/preferences/dialog/dlgprefmixer.cpp:87]: even more so if release applies **all** pending changes.

Comment (Swiftb0y) [reply src/preferences/dialog/dlgprefmixer.cpp:87]: Yes.

File: src/preferences/dialog/dlgprefmixer.h (hunk 1 above)
Comment (Swiftb0y) [reply src/preferences/dialog/dlgprefmixer.h:88]: Right, I forgot how PollingControlProxy actually works (I thought it emulates signals and slots by continuously polling a CO).

**Comment by Swiftb0y** [comment]
My manual test concluded that "Bypass EQ effect processing" is not working. The rest of the settings on that page is fine.

**Comment by ronso0** [comment]
> My manual test concluded that "Bypass EQ effect processing" is not working.

Whoopsie! Wrong cast of checked state, fixed.

**Comment by Swiftb0y** [comment]
Thank you, thats fixed. Do you just want to get the merged or can you take care my previous suggestions?

**Comment by ronso0** [comment]
One thing at a time. I'd appreciate to get this merged first. Then me or someone else can fix the UX inconsistencies and tweak the code (I already mentioned #4667, anyone could pick that up, it's just rather boring work )

**Review by Swiftb0y — APPROVED** [review]
Thank you

**Pull request merged**